src/deadline/job_attachments/api/attachment.py

```diff
@@ -28,56 +33,156 @@ def attachment_download(
+    s3_settings: JobAttachmentS3Settings = JobAttachmentS3Settings.from_s3_root_uri(s3_root_uri)
+    download_summary: DownloadSummaryStatistics = download_files_from_manifests(
+        s3_bucket=s3_settings.s3BucketName,
+        manifests_by_root=merged_manifests_by_root,
+        cas_prefix=s3_settings.full_cas_prefix(),
+        session=boto3_session,
+    )
+    logger.echo(download_summary)
+    logger.json(asdict(download_summary.convert_to_summary_statistics()))
+
+
+def attachment_upload(
+    manifests: List[str],
+    s3_root_uri: str,
+    boto3_session: boto3.Session,
+    root_dirs: List[str] = [],
+    path_mapping_rules: Optional[str] = None,
+    upload_manifest_path: Optional[str] = None,
+    logger: ClickLogger = ClickLogger(False),
+):
+    """
+    BETA API - This API is still evolving but will be made public in the near future.
+
+    API to download job attachments based on given list of manifests.
+    If path mapping rules file is given, map to corresponding destinations.
+
+    Args:
+        manifests (List[str]): File Path to the manifest file for upload.
+        s3_root_uri (str): S3 root uri including bucket name and root prefix.
+        boto3_session (boto3.Session): Boto3 session for interacting with customer s3.
+        root_dirs (List[str]): List of root directories holding attachments. Defaults to empty.
+        path_mapping_rules (Optional[str], optional): Optional file path to a JSON file contains list of path mapping. Defaults to None.
+        upload_manifest_path (Optional[str], optional): Optional prefix for uploading given manifests. Defaults to None.
+        logger (ClickLogger, optional): Logger to provide visibility. Defaults to ClickLogger(False).
+
+    Raises:
+        NonValidInputError: _description_
+    """
+
+    if not all(os.path.isdir(root) for root in root_dirs):
+        raise NonValidInputError(f"Specified root dir {root_dirs} contains nonvalid directory.")
+
+    file_name_manifest_dict: Dict[str, BaseAssetManifest] = _get_file_name_manifest_dict(
+        manifests=manifests
+    )
+    path_mapping_rule_list: List[PathMappingRule] = _process_path_mapping(
+        path_mapping_rules=path_mapping_rules, root_dirs=root_dirs
+    )
+
+    s3_settings: JobAttachmentS3Settings = JobAttachmentS3Settings.from_s3_root_uri(s3_root_uri)
+    asset_uploader: S3AssetUploader = S3AssetUploader(session=boto3_session)
+    for file_name in file_name_manifest_dict:
+        manifest: BaseAssetManifest = file_name_manifest_dict[file_name]
+
+        # File name is supposed to be prefixed by a hash of source path in path mapping or provided root dirs
+        rule: Optional[PathMappingRule] = next(
+            # search in path mapping to determine source and destination
+            (
+                rule
+                for rule in path_mapping_rule_list
+                if rule.get_hashed_source_path(manifest.get_default_hash_alg()) in file_name
```

Comment:
This implies the file name of the manifest is always going to contain the hash. 

If we are doing this, we should also change snapshot to output the file name as ```{user-supplied-name}-{hash}-{timestamp}``` This will allow us to obfuscate the use of the hash to the root dir transparently. Right now, snapshot will NOT be easy to use with this attachment upload.

I can make this change in the morning to snapshot if we agree.

Reply:
This is symmetrical to how `attachment upload` maps path mapping rule to its manifest.

I think we would need the hash in the file name, otherwise the download input data cannot maintain this correlation. The caller could pass in the hash as supplied name, but I think having the hash explicitly in the name is good for consistency. This is important for the snapshot generated manifest files being used by upload and download out of the box.